{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Guide Calling Using the Cellranger Module HDF5 File \n",
    "## Utilising Direct Capture Perturb-Seq to Deciphering Alternative Promoter Usage  \n",
    "\n",
    "1. Run through the 10x Cellranger pipeline and velocyto for single cell RNAseq quatification and using (2) guides quantifiction. all found in the cellranger files folder **bash**\n",
    "2.  Guide Calling for dual guide. Use repogle method to take molecule.h5 generated by cellranger and py to run through repogle version of guide calling or use cellranger_guidecalling.ipynb for Direct Capture Perturb-Seq dual guide. Formed guide-specific lists of cells.\n",
    "3. Pseudobulk analysis.\n",
    "    A. Seperation of guide-specific fastq files. **bash**\n",
    "    B. Whippet pseudobulk for transcript specific analysis, post UMI deduplication.  **bash**\n",
    "    C. Transcript quality control. **R**\n",
    "    D. Whippet result visualisation.\n",
    "4. Normalisation of adata object and E-distance of KD\n",
    "5. Check gene and neighboring gene expression\n",
    "6. Create individual umaps per gene of interest\n",
    "    A. UMAPs \n",
    "    B. Rand Index score\n",
    "7. Cell phase assignment model from FUCCI-matched single cell paper (GSE146773)\n",
    "8. **Differential Expression analysis.**\n",
    "    A. Find the shared P1 and P2 genes. \n",
    "    B. Check the shared P1 and P2 across different protospacers with the same A/B and C/D.\n",
    "9.  CNV Score & Numbat to quantify and Velocity quantification with loom file\n",
    "10. ESR1-specific analysis from proliferation analysis to rt-qpcr\n",
    "11.  Spectra analysis and visualisation for pathway enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scanpy 1.10.3\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%matplotlib inline\n",
    "%autoreload 2\n",
    "#general\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as pl\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import hdf5plugin\n",
    "\n",
    "#form a location\n",
    "loc=\"/Users/helenking/Desktop/Weatheritt_Lab_Y2/alt-prom-crispr-fiveprime/\"\n",
    "\n",
    "import seaborn as sns\n",
    "from tqdm.notebook import tqdm\n",
    "import scperturb\n",
    "import sys\n",
    "sys.path.append(loc+'scripts/')\n",
    "from apu_analysis import *\n",
    "import scperturb\n",
    "import infercnvpy as cnv\n",
    "from apu_analysis.cell_import import CellPopulation\n",
    "from IPython.display import clear_output\n",
    "pd.options.display.float_format = '{:.4f}'.format\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#for this python\n",
    "from scipy.special import rel_entr\n",
    "import sklearn.cluster as cluster\n",
    "import umap\n",
    "from scipy import stats\n",
    "from scipy.stats import bootstrap\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from numpy import reshape\n",
    "from numpy import array\n",
    "from sklearn.decomposition import PCA\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import ttest_ind\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "# Taken from: \n",
    "# Adamson, B.A., Norman, T.M., *et al.* \"A multiplexed CRISPR screening platform enables systematic dissection of the unfolded protein response\", *Cell*, 2016.\n",
    "# My experiment deals with two KDs- one of the MP, one of the AP using two guides. Positve controls include GINS1 ect. This is a combnatorial KD double for the same gene. No treatments were used\n",
    "\n",
    "# colours using garvan \n",
    "color1 ='#4d00c7'\n",
    "palecolor1=\"#b366ff\"\n",
    "color2= '#da3c07'\n",
    "palecolor2=\"#ff8954\"\n",
    "color3='#05d3d3'\n",
    "color4='#c6c7c5'\n",
    "color4=\"#434541\"\n",
    "color5=\"#eb31e1\"\n",
    "color6=\"#3175eb\"\n",
    "color7=\"#a7eb31\"\n",
    "color8=\"#b366ff\"\n",
    "color9=\"#ff8954\"\n",
    "color10=\"#35c9d4\"\n",
    "\n",
    "# #use viridis\n",
    "# color1=\"#fde725\"\n",
    "# color2=\"#7ad151\"\n",
    "# color3=\"#22a884\"\n",
    "# color4=\"#2a788e\"\n",
    "# color5=\"#2a788e\"\n",
    "# color6='#440154'\n",
    "# %%capture\n",
    "\n",
    "\n",
    "# Create the color palette\n",
    "palette = sns.color_palette([palecolor1,palecolor2,color3,color4])\n",
    "palette2 = sns.color_palette([color1, color2, palecolor1, palecolor2,color5,color6 ,color7])\n",
    "\n",
    "# # Create the color palette\n",
    "# palette = sns.color_palette([color1, color2,color3])\n",
    "new_palette = sns.color_palette([color1, color2,color1, color2,color1, color2,color1, color2,color1, color2,color1, color2, color3, color4])\n",
    "\n",
    "from random import randrange\n",
    "pvalue=0.05\n",
    "\n",
    "print(\"Scanpy\", sc.__version__)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "adata = ad.read_h5ad(loc+\"files/adata_normalised.h5ad\")\n",
    "#set the directio\n",
    "#perform log1p on the data\n",
    "adata.obs[\"guide_assignment_simple\"]=np.where(((adata.obs[\"guide_assignment\"].str.startswith(\"non\"))|(adata.obs[\"guide_assignment\"].str.startswith(\"Non\"))), \"non-targeting\", adata.obs[\"guide_assignment\"])\n",
    "gene_list_select=adata.obs[\"guide_assignment_simple\"][adata.obs[\"guide_assignment_simple\"]!=\"non-targeting\"].drop_duplicates().tolist()\n",
    "%%capture\n",
    "# sc.tl.rank_genes_groups(adata, 'guide_assignment_simple', groups=gene_list_select,reference=\"non-targeting\", method='wilcoxon', use_raw=False)\n",
    "sc.tl.rank_genes_groups(adata, 'guide_assignment_simple', groups=gene_list_select,reference=\"non-targeting\", method='t-test', use_raw=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the differentially expressed genes\n",
    "# %%capture \n",
    "gene_list=[]\n",
    "pvalue=0.05\n",
    "number=1000\n",
    "sortby=\"pvals_adj_1\"\n",
    "min_fc=1\n",
    "#get the highly_variable genes\n",
    "extract_highly_variable_genes=adata.var[adata.var[\"highly_variable\"]==True].index\n",
    "gene_list_recursive=adata.obs[\"perturbation\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values\n",
    "#remove ARNT and if gene_A CSNK1E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the differentially expressed genes\n",
    "# %%capture \n",
    "gene_list=[]\n",
    "pvalue=0.05\n",
    "logFC=None\n",
    "number=1000\n",
    "sortby=\"rank_1\"\n",
    "min_fc=0.5\n",
    "#get the highly_variable genes\n",
    "extract_highly_variable_genes=adata.var[adata.var[\"highly_variable\"]==True].index\n",
    "gene_list_recursive=adata.obs[\"perturbation\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values\n",
    "#remove ARNT and if gene_A CSNK1E\n",
    "gene_list_recursive=gene_list_recursive[~gene_list_recursive.isin([\"ARNT\",\"CSNK1E\"])]\n",
    "for gene in adata.obs[\"perturbation\"][(adata.obs[\"guide_id\"]!=\"non-targeting_Control\") & (adata.obs[\"successfulKD\"]==\"True\")].drop_duplicates().values:\n",
    "    # print(gene)\n",
    "    #AP\n",
    "    gene_AP=gene+\"_AP\"\n",
    "    if gene_AP not in adata.obs[\"guide_id\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values:\n",
    "        continue\n",
    "    elif gene in [\"ARNT\"]:\n",
    "        print(\"ARNT\")\n",
    "    else:\n",
    "        for promoter in [\"MP\",\"AP\"]:\n",
    "            gene_promoter=gene+\"_\"+promoter+\"_\"+str(1)\n",
    "            gene_promoter_2=gene+\"_\"+promoter+\"_\"+str(2)\n",
    "            gene_promoter_random_1=gene_list_recursive[randrange(len(gene_list_recursive))]+\"_\"+promoter+\"_\"+str(1)\n",
    "            gene_promoter_random_2=gene_list_recursive[randrange(len(gene_list_recursive))]+\"_\"+promoter+\"_\"+str(2)\n",
    "            ##form a dataframe with custom colummns \n",
    "            #subset random gene from gene list\n",
    "            protospacer_1= sc.get.rank_genes_groups_df(adata, group=gene_promoter,pval_cutoff=pvalue)\n",
    "            #order by the absolute value of the logfoldchange\n",
    "            protospacer_1[\"abs_logfoldchanges\"]=protospacer_1[\"logfoldchanges\"].abs()\n",
    "            protospacer_1.sort_values(by=\"abs_logfoldchanges\",ascending=False,inplace=True)\n",
    "            #add a rank column\n",
    "            protospacer_1[\"rank\"]=range(1,len(protospacer_1)+1)\n",
    "            protospacer_2= sc.get.rank_genes_groups_df(adata, group=gene_promoter_2,pval_cutoff=pvalue)\n",
    "            protospacer_2[\"abs_logfoldchanges\"]=protospacer_2[\"logfoldchanges\"].abs()\n",
    "            protospacer_2.sort_values(by=\"abs_logfoldchanges\",ascending=False,inplace=True)\n",
    "            #add a rank column\n",
    "            protospacer_2[\"rank\"]=range(1,len(protospacer_2)+1)\n",
    "            protospacer_1_random= sc.get.rank_genes_groups_df(adata, group=gene_promoter_random_1,pval_cutoff=pvalue)\n",
    "            protospacer_1_random[\"abs_logfoldchanges\"]=protospacer_1_random[\"logfoldchanges\"].abs()\n",
    "            protospacer_1_random.sort_values(by=\"abs_logfoldchanges\",ascending=False,inplace=True)\n",
    "            #add a rank column\n",
    "            protospacer_1_random[\"rank\"]=range(1,len(protospacer_1_random)+1)\n",
    "            protospacer_2_random= sc.get.rank_genes_groups_df(adata, group=gene_promoter_random_2,pval_cutoff=pvalue)\n",
    "            protospacer_2_random[\"abs_logfoldchanges\"]=protospacer_2_random[\"logfoldchanges\"].abs()\n",
    "            protospacer_2_random.sort_values(by=\"abs_logfoldchanges\",ascending=False,inplace=True)\n",
    "            #add a rank column\n",
    "            protospacer_2_random[\"rank\"]=range(1,len(protospacer_2_random)+1)\n",
    "            gene_df=protospacer_1.merge(protospacer_2, on=\"names\", how=\"left\", suffixes=('_1', '_2')).dropna()\n",
    "            #filter for highly variable genes\n",
    "            gene_df_random_1=protospacer_1.merge(protospacer_1_random, on=\"names\", how=\"left\", suffixes=('_1', '_2')).dropna()\n",
    "            gene_df_random_2=protospacer_1.merge(protospacer_2_random, on=\"names\", how=\"left\", suffixes=('_1', '_2')).dropna()\n",
    "            rank_columns=[\"rank_1\",\"rank_2\"]\n",
    "            #get the correlation for  scores_1 and scores_2 for corr\n",
    "            correlation_1=gene_df.sort_values(by=sortby, ascending=True)[rank_columns].corr(method='spearman',numeric_only=True)\n",
    "            correlation_random_1=gene_df_random_1.sort_values(by=sortby, ascending=True)[rank_columns].corr(method='spearman',numeric_only=True)\n",
    "            correlation_random_2=gene_df_random_2.sort_values(by=sortby, ascending=True)[rank_columns].corr(method='spearman',numeric_only=True)\n",
    "            #create a row wtht he gene name \n",
    "            gene_list_row=[gene,promoter]\n",
    "            gene_list_row.extend([correlation_1.loc[\"rank_1\",\"rank_2\"],correlation_random_1.loc[\"rank_1\",\"rank_2\"],correlation_random_2.loc[\"rank_1\",\"rank_2\"]])\n",
    "            gene_list.append(gene_list_row)\n",
    "            \n",
    "\n",
    "#flatten list of lists into dataframe\n",
    "gene_df_final=pd.DataFrame(gene_list, columns=[\"gene\",\"promoter\",\"scores_1_targeted\",\"scores_1_random\",\"scores_2_random\"])     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AHCYL1\n",
      "SP1\n",
      "TGIF1\n",
      "YWHAZ\n",
      "P4HB\n",
      "NCOR2\n",
      "JARID2\n",
      "ESR1\n",
      "PA2G4\n",
      "GPBP1\n",
      "DHX30\n",
      "SAFB\n",
      "NFE2L2\n",
      "CHD8\n",
      "SIN3A\n",
      "PSMC5\n",
      "RC3H2\n",
      "ZNF3\n",
      "GTF2F1\n",
      "ADAR\n",
      "NR2F2\n",
      "BZW1\n",
      "FHL2\n",
      "SET\n",
      "MYBBP1A\n",
      "CUX1\n",
      "NBN\n",
      "CALR\n",
      "APP\n",
      "STAG2\n",
      "PKM\n",
      "BRIP1\n",
      "GATA3\n",
      "RBM47\n"
     ]
    }
   ],
   "source": [
    "#get the differentially expressed genes\n",
    "# %%capture \n",
    "# pvalue=0.01\n",
    "gene_list=[]\n",
    "gene_list_recursive=adata.obs[\"perturbation\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values\n",
    "#remove ARNT and if gene_A CSNK1E\n",
    "gene_list_recursive=gene_list_recursive[~gene_list_recursive.isin([\"ARNT\",\"CSNK1E\"])]\n",
    "for gene in adata.obs[\"perturbation\"][(adata.obs[\"guide_id\"]!=\"non-targeting_Control\") & (adata.obs[\"successfulKD\"]==\"True\")].drop_duplicates().values:\n",
    "    print(gene)\n",
    "    #AP\n",
    "    gene_AP=gene+\"_AP\"\n",
    "    if gene_AP not in adata.obs[\"guide_id\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values:\n",
    "        continue\n",
    "    elif gene in [\"ARNT\"]:\n",
    "        print(\"ARNT\")\n",
    "    else:\n",
    "        for promoter in [\"MP\",\"AP\"]:\n",
    "            gene_promoter=gene+\"_\"+promoter+\"_\"+str(1)\n",
    "            protospacer_1= sc.get.rank_genes_groups_df(adata, group=gene_promoter,pval_cutoff=pvalue)\n",
    "            protospacer_1=protospacer_1['names'].values\n",
    "            \n",
    "            gene_promoter=gene+\"_\"+promoter+\"_\"+str(2)\n",
    "            protospacer_2= sc.get.rank_genes_groups_df(adata, group=gene_promoter,pval_cutoff=pvalue)\n",
    "            protospacer_2=protospacer_2['names'].values\n",
    "            \n",
    "            gene_promoter=gene_list_recursive[randrange(len(gene_list_recursive))]+\"_\"+promoter+\"_\"+str(1)\n",
    "            protospacer_1_random= sc.get.rank_genes_groups_df(adata, group=gene_promoter,pval_cutoff=pvalue)\n",
    "            protospacer_1_random=protospacer_1_random[\"names\"].values\n",
    "\n",
    "            gene_promoter=gene_list_recursive[randrange(len(gene_list_recursive))]+\"_\"+promoter+\"_\"+str(2)\n",
    "            protospacer_2_random= sc.get.rank_genes_groups_df(adata, group=gene_promoter,pval_cutoff=pvalue)\n",
    "            protospacer_2_random=protospacer_2_random[\"names\"].values\n",
    "            \n",
    "            \n",
    "            overlap = [x for x in protospacer_1 if x in protospacer_2]\n",
    "            overlap_random_1 = [x for x in overlap if x in protospacer_1_random]\n",
    "            overlap_random_2 = [x for x in overlap if x in protospacer_2_random]\n",
    "            #if overlap protospacer_1 protospacer_2 overlap_random_2 overlap_random_1 = 0 set to 1\n",
    "            # print(protospacer_1,protospacer_2,overlap,overlap_random_1,overlap_random_2)\n",
    "            try: \n",
    "                overlap_1=len(overlap)/len(protospacer_1)\n",
    "                overlap_2=len(overlap)/len(protospacer_2)\n",
    "                overlap_random_1=len(overlap_random_1)/len(protospacer_1_random)\n",
    "                overlap_random_2=len(overlap_random_2)/len(protospacer_2_random)\n",
    "            except ZeroDivisionError:\n",
    "                overlap_1=0\n",
    "                overlap_2=0\n",
    "                overlap_random_1=0\n",
    "                overlap_random_2=0\n",
    "            \n",
    "            gene_df=pd.DataFrame([[gene,promoter,overlap_1,overlap_2,overlap_random_1,overlap_random_2]])\n",
    "            # gene_df_details=pd.DataFrame([[gene,promoter,protospacer_1,protospacer_2,overlap,len(protospacer_1),len(protospacer_2),len(overlap)]])\n",
    "\n",
    "            # print(gene_df.shape)\n",
    "            gene_list.append(gene_df)\n",
    "            # break\n",
    "\n",
    "\n",
    "# col_names=[\"Gene\",\"Promoter\",\"Protospacer_1\",\"Protospacer_2\",\"Overlap\",\"Protospacer_Random_1\",\"Protospacer_Random_2\"]\n",
    "#add col_names\n",
    "col_names=[\"Gene\",\"Promoter\",\"Overlap_1\",\"Overlap_2\",\"Overlap_Random_1\",\"Overlap_Random_2\"]\n",
    "all=pd.DataFrame(pd.concat(gene_list))\n",
    "all.columns=col_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TtestResult(statistic=-1.200758436011414, pvalue=0.23196364293947955, df=133.97197243239725)\n",
      "TtestResult(statistic=0.14299988815177855, pvalue=0.8865064991701896, df=133.1170801399246)\n",
      "TtestResult(statistic=10.31768853968277, pvalue=6.501330271631454e-18, df=111.19538835724522)\n",
      "TtestResult(statistic=11.997174432050567, pvalue=2.1194060394826254e-21, df=104.97349430144247)\n"
     ]
    }
   ],
   "source": [
    "#get the tests between two unpaired parametric groups\n",
    "#melt for random\n",
    "all_melt=all.melt(id_vars=[\"Gene\",\"Promoter\"])\n",
    "all.index=all[\"Gene\"]+\"_\"+all[\"Promoter\"]   \n",
    "print(stats.ttest_ind(all[\"Overlap_1\"],all[\"Overlap_2\"],equal_var=False))\n",
    "print(stats.ttest_ind(all[\"Overlap_Random_1\"],all[\"Overlap_Random_2\"],equal_var=False))\n",
    "print(stats.ttest_ind(all[\"Overlap_1\"],all[\"Overlap_Random_1\"],equal_var=False))\n",
    "print(stats.ttest_ind(all[\"Overlap_2\"],all[\"Overlap_Random_2\"],equal_var=False))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot the overlap between the two guides\n",
    "sns.set(style=\"white\")\n",
    "sns.set_context(\"talk\")\n",
    "fig, ax = plt.subplots(figsize=(8, 5))\n",
    "new_pallete=sns.color_palette([\"#B4D88C\",\"#C67B6A\",\"#53449B\",\"#8C1F1C\"])\n",
    "#show the percentage over \"Overlap_length\"/\"Protospacer_1_length\"\n",
    "sns.kdeplot(data=all_melt,x=\"value\",hue=\"variable\",palette=new_pallete,  \n",
    ")\n",
    "#save plot\n",
    "plt.savefig(loc+\"figures/overlap_gene_signature.pdf\")\n",
    "gene_list_select=adata.obs[\"guide_id\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values\n",
    "sc.tl.rank_genes_groups(adata, 'guide_id', groups=gene_list_select.to_list(),reference=\"non-targeting_Control\", method='t-test')\n",
    "genelist=adata.obs['guide_id'].drop_duplicates().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "sc.tl.rank_genes_groups(adata, 'guide_id',groups=genelist, reference=\"non-targeting_Control\", use_raw=False)\n",
    "\n",
    "#get target gene logfoldchanges\n",
    "#get the differentially expressed genes \n",
    "gene_list=[]\n",
    "comp_gene_list=[]\n",
    "\n",
    "for gene in adata.obs[\"perturbation\"][(adata.obs[\"guide_id\"]!=\"non-targeting_Control\")& (adata.obs[\"successfulKD\"]==\"True\")].drop_duplicates().values:\n",
    "    # print(gene)\n",
    "    #AP\n",
    "    gene_AP=gene+\"_AP\"\n",
    "    gene_MP=gene+\"_MP\"\n",
    "\n",
    "    if gene_AP not in adata.obs[\"guide_id\"][(adata.obs[\"guide_id\"]!=\"non-targeting_Control\")& (adata.obs[\"successfulKD\"]==\"True\")].drop_duplicates().values:\n",
    "        continue\n",
    "    else:\n",
    "        ##form a dataframe with custom colummns \n",
    "        MP = sc.get.rank_genes_groups_df(adata, group=gene_MP)\n",
    "        MP=MP[(MP[\"names\"]==gene)]\n",
    "        AP = sc.get.rank_genes_groups_df(adata, group=gene_AP)        \n",
    "        AP=AP[(AP[\"names\"]==gene)]\n",
    "        promoter_line=[gene,MP[\"scores\"].values[0],AP[\"scores\"].values[0],MP[\"pvals_adj\"].values[0],AP[\"pvals_adj\"].values[0]]\n",
    "        comp_gene_list.append(promoter_line)\n",
    "        # break\n",
    "#convert to dataframe\n",
    "comp_gene_list=pd.DataFrame(comp_gene_list, columns=[\"gene\",\"MP_logfoldchange\",\"AP_logfoldchange\",\"MP_pvalue\",\"AP_pvalue\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "gene_list=[]\n",
    "comp_gene_list=[]\n",
    "\n",
    "for gene in adata.obs[\"perturbation\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values:\n",
    "    print(gene)\n",
    "    #AP\n",
    "    gene_AP=gene+\"_AP\"\n",
    "    gene_MP=gene+\"_MP\"\n",
    "\n",
    "    if gene_AP not in adata.obs[\"guide_id\"][adata.obs[\"guide_id\"]!=\"non-targeting_Control\"].drop_duplicates().values:\n",
    "        continue\n",
    "    else:\n",
    "        ##form a dataframe with custom colummns \n",
    "        MP = sc.get.rank_genes_groups_df(adata, group=gene_MP, pval_cutoff=0.05)\n",
    "        # MP = MP[MP[\"logfoldchanges\"].abs()>0.1]\n",
    "        MP_names=MP['names'].values\n",
    "        # ,   pval_cutoff=0.05)['names']\n",
    "        # sc.tl.rank_genes_groups(adata_ideal_subset, 'guide_identity', groups=['MYBBP1A_AP'], reference='Negative_Control_Control', method='wilcoxon')\n",
    "        AP = sc.get.rank_genes_groups_df(adata, group=gene_AP, pval_cutoff=0.05)\n",
    "        # AP = AP[AP[\"logfoldchanges\"].abs()>0.5]\n",
    "\n",
    "        AP_names=AP['names'].values\n",
    "\n",
    "        # , pval_cutoff=0.05, log2fc_min=0)['names']\n",
    "        overlap = [x for x in MP_names if x in AP_names]\n",
    "        #get average log2fc for MP, AP and overlap\n",
    "        \n",
    "        gene_df=pd.DataFrame([[gene,MP_names,AP_names,overlap,len(MP_names),len(AP_names),len(overlap)]])\n",
    "        print(gene_df.shape)\n",
    "        gene_list.append(gene_df)\n",
    "\n",
    "\n",
    "        ##add to another df which includes the gene and their direction\n",
    "        MP[\"Guide\"]=gene_MP\n",
    "        comp_gene_list.append(MP)\n",
    "        AP[\"Guide\"]=gene_AP\n",
    "        comp_gene_list.append(AP)\n",
    "        # break\n",
    "\n",
    "col_names=[\"Target_Gene\",\"MP_Genes\",\"AP_Genes\",\"Overlap_Genes\",\"MP_Gene_Num\",\"AP_Gene_Num\",\"Overlap_Gene_Num\"]\n",
    "all=pd.DataFrame(pd.concat(gene_list))\n",
    "all.columns=col_names\n",
    "\n",
    "#only filter for genes with successfulKD \n",
    "all_filt=all[all[\"Target_Gene\"].isin(adata.obs[\"gene\"][adata.obs[\"successfulKD\"]==\"True\"].drop_duplicates().values)]\n",
    "all_filt.to_csv(loc+\"files/singlecell_shortread_analysis/differential_exp_list.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_comp=pd.DataFrame(pd.concat(comp_gene_list))\n",
    "all_comp.columns=MP.columns\n",
    "all_comp[\"Promoter\"]=all_comp[\"Guide\"].str.split(\"_\").str.get(1)\n",
    "all_comp[\"Target_Gene\"]=all_comp[\"Guide\"].str.split(\"_\").str.get(0)\n",
    "all_comp_mean=all_comp.groupby(\"Target_Gene\").agg({'logfoldchanges': 'mean'})\n",
    "all_comp_mean.reset_index(inplace=True)\n",
    "all_comp_mean.columns=[\"Target_Gene\",\"logfoldchanges_mean\"]\n",
    "all_comp=all_comp.merge(all_comp_mean,on=\"Target_Gene\")\n",
    "all_comp_mean=all_comp.groupby(\"Target_Gene\").agg({'Target_Gene':\"count\"}).rename(columns={\"Target_Gene\":\"Count\"})\n",
    "all_comp_mean.reset_index(inplace=True)\n",
    "all_comp=all_comp.merge(all_comp_mean,on=\"Target_Gene\")\n",
    "all_comp=all_comp.sort_values(by=\"Guide\")\n",
    "all_comp.to_csv(loc+\"files/singlecell_shortread_analysis/differential_exp_list_full.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "all[\"Mutual_Exclusivity\"]=all[\"Overlap_Gene_Num\"]/(all[\"MP_Gene_Num\"]+all[\"AP_Gene_Num\"])\n",
    "\n",
    "n_term=pd.read_table(\"alt-prom-crispr-fiveprime/reference/all_pivot_simple_nterm.txt\")\n",
    "n_term=all.merge(n_term, right_on=\"Gene_symbol\",left_on=\"Target_Gene\")\n",
    "\n",
    "n_term=n_term[[\"MP_Gene_Num\",\"AP_Gene_Num\",\"Overlap_Gene_Num\",\"Mutual_Exclusivity\",\"Ensembl_ID\",\"Gene_symbol\",\"Nterminus_Change\"]]\n",
    "# all_comp_pivot\n",
    "all_comp_all=all_comp.merge(n_term, left_on=\"Target_Gene\", right_on=\"Gene_symbol\" ,how=\"outer\")\n",
    "all_comp_all=all_comp_all.sort_values(\"Nterminus_Change\",ascending=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "apu",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
