

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Guide Calling Using the Cellranger Module HDF5 File &mdash; Isoform Specific Perturb-Seq  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../_static/css/theme.css?v=9edc463e" />
      <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../../_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../_static/documentation_options.js?v=5929fcd5"></script>
      <script src="../../_static/doctools.js?v=fd6eb6e6"></script>
      <script src="../../_static/sphinx_highlight.js?v=6ffebe34"></script>
      <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="../../_static/copybutton.js?v=f281be69"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@4/tex-mml-chtml.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Guide Calling Using the Cellranger Module HDF5 File" href="9A_cnv_score.html" />
    <link rel="prev" title="Guide Calling Using the Cellranger Module HDF5 File" href="7A_cellphase_model.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../index.html" class="icon icon-home">
            Isoform Specific Perturb-Seq
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Analysis Pipeline:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../promoter_identification.html">Promoter Identification</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../guide_design.html">Guide Design</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#preprocessing-guide-calling">Preprocessing &amp; Guide Calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#transcriptional-divergence-e-stats">Transcriptional Divergence (E-Stats)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analysis_scripts.html#cell-phase-pathway-modeling">Cell Phase &amp; Pathway Modeling</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../analysis_scripts.html#clinical-differential-expression">Clinical &amp; Differential Expression</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="2A_cellranger_guidecalling.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="2B_guide_calling_cellrangerfilter4_idealguides.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="4_estatistic_gene_guide.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="5_genekd_neighbouring_gene_expression.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="6_perturbseq_tsne_kldivergence_umap.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="7A_cellphase_model.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Guide Calling Using the Cellranger Module HDF5 File</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Utilising-Direct-Capture-Perturb-Seq-to-Deciphering-Alternative-Promoter-Usage">Utilising Direct Capture Perturb-Seq to Deciphering Alternative Promoter Usage</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="9A_cnv_score.html">Guide Calling Using the Cellranger Module HDF5 File</a></li>
<li class="toctree-l3"><a class="reference internal" href="10D_esr1_survival_curve.html">GEPIA2 Run Through</a></li>
<li class="toctree-l3"><a class="reference internal" href="11A_spectra.html">Spectra Analysis</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Isoform Specific Perturb-Seq</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../analysis_scripts.html">Analysis Scripts &amp; Pipelines</a></li>
      <li class="breadcrumb-item active">Guide Calling Using the Cellranger Module HDF5 File</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/scripts/3_analysis_scripts/8_perturbseq_differentialexpression.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Guide-Calling-Using-the-Cellranger-Module-HDF5-File">
<h1>Guide Calling Using the Cellranger Module HDF5 File<a class="headerlink" href="#Guide-Calling-Using-the-Cellranger-Module-HDF5-File" title="Link to this heading"></a></h1>
<section id="Utilising-Direct-Capture-Perturb-Seq-to-Deciphering-Alternative-Promoter-Usage">
<h2>Utilising Direct Capture Perturb-Seq to Deciphering Alternative Promoter Usage<a class="headerlink" href="#Utilising-Direct-Capture-Perturb-Seq-to-Deciphering-Alternative-Promoter-Usage" title="Link to this heading"></a></h2>
<ol class="arabic simple">
<li><p>Run through the 10x Cellranger pipeline and velocyto for single cell RNAseq quatification and using (2) guides quantifiction. all found in the cellranger files folder <strong>bash</strong></p></li>
<li><p>Guide Calling for dual guide. Use repogle method to take molecule.h5 generated by cellranger and py to run through repogle version of guide calling or use cellranger_guidecalling.ipynb for Direct Capture Perturb-Seq dual guide. Formed guide-specific lists of cells.</p></li>
<li><p>Pseudobulk analysis. A. Seperation of guide-specific fastq files. <strong>bash</strong> B. Whippet pseudobulk for transcript specific analysis, post UMI deduplication. <strong>bash</strong> C. Transcript quality control. <strong>R</strong> D. Whippet result visualisation.</p></li>
<li><p>Normalisation of adata object and E-distance of KD</p></li>
<li><p>Check gene and neighboring gene expression</p></li>
<li><p>Create individual umaps per gene of interest A. UMAPs B. Rand Index score</p></li>
<li><p>Cell phase assignment model from FUCCI-matched single cell paper (GSE146773)</p></li>
<li><p><strong>Differential Expression analysis.</strong> A. Find the shared P1 and P2 genes. B. Check the shared P1 and P2 across different protospacers with the same A/B and C/D.</p></li>
<li><p>CNV Score &amp; Numbat to quantify and Velocity quantification with loom file</p></li>
<li><p>ESR1-specific analysis from proliferation analysis to rt-qpcr</p></li>
<li><p>Spectra analysis and visualisation for pathway enrichment</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">load_ext</span> autoreload
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="o">%</span><span class="k">autoreload</span> 2
<span class="c1">#general</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pl</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">hdf5plugin</span>

<span class="c1">#form a location</span>
<span class="n">loc</span><span class="o">=</span><span class="s2">&quot;/Users/helenking/Desktop/Weatheritt_Lab_Y2/alt-prom-crispr-fiveprime/&quot;</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm.notebook</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scperturb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s1">&#39;scripts/&#39;</span><span class="p">)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis</span><span class="w"> </span><span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scperturb</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">infercnvpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cnv</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">apu_analysis.cell_import</span><span class="w"> </span><span class="kn">import</span> <span class="n">CellPopulation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">IPython.display</span><span class="w"> </span><span class="kn">import</span> <span class="n">clear_output</span>
<span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">display</span><span class="o">.</span><span class="n">float_format</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{:.4f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>

<span class="c1">#for this python</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.special</span><span class="w"> </span><span class="kn">import</span> <span class="n">rel_entr</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sklearn.cluster</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">cluster</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">umap</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">bootstrap</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">statsmodels.api</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sm</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">statsmodels.stats.multitest</span><span class="w"> </span><span class="kn">import</span> <span class="n">multipletests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">reshape</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="kn">import</span> <span class="n">array</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.decomposition</span><span class="w"> </span><span class="kn">import</span> <span class="n">PCA</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">stats</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">ttest_ind</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">statsmodels.stats.multitest</span><span class="w"> </span><span class="kn">import</span> <span class="n">multipletests</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="kn">import</span> <span class="n">mannwhitneyu</span>

<span class="c1"># Taken from:</span>
<span class="c1"># Adamson, B.A., Norman, T.M., *et al.* &quot;A multiplexed CRISPR screening platform enables systematic dissection of the unfolded protein response&quot;, *Cell*, 2016.</span>
<span class="c1"># My experiment deals with two KDs- one of the MP, one of the AP using two guides. Positve controls include GINS1 ect. This is a combnatorial KD double for the same gene. No treatments were used</span>

<span class="c1"># colours using garvan</span>
<span class="n">color1</span> <span class="o">=</span><span class="s1">&#39;#4d00c7&#39;</span>
<span class="n">palecolor1</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color2</span><span class="o">=</span> <span class="s1">&#39;#da3c07&#39;</span>
<span class="n">palecolor2</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color3</span><span class="o">=</span><span class="s1">&#39;#05d3d3&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s1">&#39;#c6c7c5&#39;</span>
<span class="n">color4</span><span class="o">=</span><span class="s2">&quot;#434541&quot;</span>
<span class="n">color5</span><span class="o">=</span><span class="s2">&quot;#eb31e1&quot;</span>
<span class="n">color6</span><span class="o">=</span><span class="s2">&quot;#3175eb&quot;</span>
<span class="n">color7</span><span class="o">=</span><span class="s2">&quot;#a7eb31&quot;</span>
<span class="n">color8</span><span class="o">=</span><span class="s2">&quot;#b366ff&quot;</span>
<span class="n">color9</span><span class="o">=</span><span class="s2">&quot;#ff8954&quot;</span>
<span class="n">color10</span><span class="o">=</span><span class="s2">&quot;#35c9d4&quot;</span>

<span class="c1"># #use viridis</span>
<span class="c1"># color1=&quot;#fde725&quot;</span>
<span class="c1"># color2=&quot;#7ad151&quot;</span>
<span class="c1"># color3=&quot;#22a884&quot;</span>
<span class="c1"># color4=&quot;#2a788e&quot;</span>
<span class="c1"># color5=&quot;#2a788e&quot;</span>
<span class="c1"># color6=&#39;#440154&#39;</span>
<span class="c1"># %%capture</span>


<span class="c1"># Create the color palette</span>
<span class="n">palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">palecolor1</span><span class="p">,</span><span class="n">palecolor2</span><span class="p">,</span><span class="n">color3</span><span class="p">,</span><span class="n">color4</span><span class="p">])</span>
<span class="n">palette2</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">palecolor1</span><span class="p">,</span> <span class="n">palecolor2</span><span class="p">,</span><span class="n">color5</span><span class="p">,</span><span class="n">color6</span> <span class="p">,</span><span class="n">color7</span><span class="p">])</span>

<span class="c1"># # Create the color palette</span>
<span class="c1"># palette = sns.color_palette([color1, color2,color3])</span>
<span class="n">new_palette</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span><span class="n">color1</span><span class="p">,</span> <span class="n">color2</span><span class="p">,</span> <span class="n">color3</span><span class="p">,</span> <span class="n">color4</span><span class="p">])</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">random</span><span class="w"> </span><span class="kn">import</span> <span class="n">randrange</span>
<span class="n">pvalue</span><span class="o">=</span><span class="mf">0.05</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Scanpy&quot;</span><span class="p">,</span> <span class="n">sc</span><span class="o">.</span><span class="n">__version__</span><span class="p">)</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/helenking/anaconda3/envs/apu/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scanpy 1.10.3
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/adata_normalised.h5ad&quot;</span><span class="p">)</span>
<span class="c1">#set the directio</span>
<span class="c1">#perform log1p on the data</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment_simple&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(((</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;non&quot;</span><span class="p">))</span><span class="o">|</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;Non&quot;</span><span class="p">))),</span> <span class="s2">&quot;non-targeting&quot;</span><span class="p">,</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment&quot;</span><span class="p">])</span>
<span class="n">gene_list_select</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment_simple&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_assignment_simple&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="o">%%</span><span class="n">capture</span>
<span class="c1"># sc.tl.rank_genes_groups(adata, &#39;guide_assignment_simple&#39;, groups=gene_list_select,reference=&quot;non-targeting&quot;, method=&#39;wilcoxon&#39;, use_raw=False)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;guide_assignment_simple&#39;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="n">gene_list_select</span><span class="p">,</span><span class="n">reference</span><span class="o">=</span><span class="s2">&quot;non-targeting&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test&#39;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get the differentially expressed genes</span>
<span class="c1"># %%capture</span>
<span class="n">gene_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">pvalue</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">number</span><span class="o">=</span><span class="mi">1000</span>
<span class="n">sortby</span><span class="o">=</span><span class="s2">&quot;pvals_adj_1&quot;</span>
<span class="n">min_fc</span><span class="o">=</span><span class="mi">1</span>
<span class="c1">#get the highly_variable genes</span>
<span class="n">extract_highly_variable_genes</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
<span class="n">gene_list_recursive</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
<span class="c1">#remove ARNT and if gene_A CSNK1E</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get the differentially expressed genes</span>
<span class="c1"># %%capture</span>
<span class="n">gene_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">pvalue</span><span class="o">=</span><span class="mf">0.05</span>
<span class="n">logFC</span><span class="o">=</span><span class="kc">None</span>
<span class="n">number</span><span class="o">=</span><span class="mi">1000</span>
<span class="n">sortby</span><span class="o">=</span><span class="s2">&quot;rank_1&quot;</span>
<span class="n">min_fc</span><span class="o">=</span><span class="mf">0.5</span>
<span class="c1">#get the highly_variable genes</span>
<span class="n">extract_highly_variable_genes</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;highly_variable&quot;</span><span class="p">]</span><span class="o">==</span><span class="kc">True</span><span class="p">]</span><span class="o">.</span><span class="n">index</span>
<span class="n">gene_list_recursive</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
<span class="c1">#remove ARNT and if gene_A CSNK1E</span>
<span class="n">gene_list_recursive</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="o">~</span><span class="n">gene_list_recursive</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;ARNT&quot;</span><span class="p">,</span><span class="s2">&quot;CSNK1E&quot;</span><span class="p">])]</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="c1"># print(gene)</span>
    <span class="c1">#AP</span>
    <span class="n">gene_AP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_AP&quot;</span>
    <span class="k">if</span> <span class="n">gene_AP</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="k">elif</span> <span class="n">gene</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ARNT&quot;</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ARNT&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span><span class="s2">&quot;AP&quot;</span><span class="p">]:</span>
            <span class="n">gene_promoter</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">gene_promoter_2</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">gene_promoter_random_1</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="n">randrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list_recursive</span><span class="p">))]</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">gene_promoter_random_2</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="n">randrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list_recursive</span><span class="p">))]</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="c1">##form a dataframe with custom colummns</span>
            <span class="c1">#subset random gene from gene list</span>
            <span class="n">protospacer_1</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="c1">#order by the absolute value of the logfoldchange</span>
            <span class="n">protospacer_1</span><span class="p">[</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">protospacer_1</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">protospacer_1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1">#add a rank column</span>
            <span class="n">protospacer_1</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_1</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">protospacer_2</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter_2</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_2</span><span class="p">[</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">protospacer_2</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">protospacer_2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1">#add a rank column</span>
            <span class="n">protospacer_2</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_2</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">protospacer_1_random</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter_random_1</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_1_random</span><span class="p">[</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">protospacer_1_random</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">protospacer_1_random</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1">#add a rank column</span>
            <span class="n">protospacer_1_random</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_1_random</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">protospacer_2_random</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter_random_2</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_2_random</span><span class="p">[</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">protospacer_2_random</span><span class="p">[</span><span class="s2">&quot;logfoldchanges&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">protospacer_2_random</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;abs_logfoldchanges&quot;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1">#add a rank column</span>
            <span class="n">protospacer_2_random</span><span class="p">[</span><span class="s2">&quot;rank&quot;</span><span class="p">]</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_2_random</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">gene_df</span><span class="o">=</span><span class="n">protospacer_1</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">protospacer_2</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_1&#39;</span><span class="p">,</span> <span class="s1">&#39;_2&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
            <span class="c1">#filter for highly variable genes</span>
            <span class="n">gene_df_random_1</span><span class="o">=</span><span class="n">protospacer_1</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">protospacer_1_random</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_1&#39;</span><span class="p">,</span> <span class="s1">&#39;_2&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
            <span class="n">gene_df_random_2</span><span class="o">=</span><span class="n">protospacer_1</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">protospacer_2_random</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;names&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_1&#39;</span><span class="p">,</span> <span class="s1">&#39;_2&#39;</span><span class="p">))</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
            <span class="n">rank_columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;rank_1&quot;</span><span class="p">,</span><span class="s2">&quot;rank_2&quot;</span><span class="p">]</span>
            <span class="c1">#get the correlation for  scores_1 and scores_2 for corr</span>
            <span class="n">correlation_1</span><span class="o">=</span><span class="n">gene_df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">sortby</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="n">rank_columns</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;spearman&#39;</span><span class="p">,</span><span class="n">numeric_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">correlation_random_1</span><span class="o">=</span><span class="n">gene_df_random_1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">sortby</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="n">rank_columns</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;spearman&#39;</span><span class="p">,</span><span class="n">numeric_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">correlation_random_2</span><span class="o">=</span><span class="n">gene_df_random_2</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">sortby</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)[</span><span class="n">rank_columns</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;spearman&#39;</span><span class="p">,</span><span class="n">numeric_only</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="c1">#create a row wtht he gene name</span>
            <span class="n">gene_list_row</span><span class="o">=</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span><span class="n">promoter</span><span class="p">]</span>
            <span class="n">gene_list_row</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">correlation_1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;rank_1&quot;</span><span class="p">,</span><span class="s2">&quot;rank_2&quot;</span><span class="p">],</span><span class="n">correlation_random_1</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;rank_1&quot;</span><span class="p">,</span><span class="s2">&quot;rank_2&quot;</span><span class="p">],</span><span class="n">correlation_random_2</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s2">&quot;rank_1&quot;</span><span class="p">,</span><span class="s2">&quot;rank_2&quot;</span><span class="p">]])</span>
            <span class="n">gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_list_row</span><span class="p">)</span>


<span class="c1">#flatten list of lists into dataframe</span>
<span class="n">gene_df_final</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span><span class="s2">&quot;promoter&quot;</span><span class="p">,</span><span class="s2">&quot;scores_1_targeted&quot;</span><span class="p">,</span><span class="s2">&quot;scores_1_random&quot;</span><span class="p">,</span><span class="s2">&quot;scores_2_random&quot;</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get the differentially expressed genes</span>
<span class="c1"># %%capture</span>
<span class="c1"># pvalue=0.01</span>
<span class="n">gene_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">gene_list_recursive</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
<span class="c1">#remove ARNT and if gene_A CSNK1E</span>
<span class="n">gene_list_recursive</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="o">~</span><span class="n">gene_list_recursive</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;ARNT&quot;</span><span class="p">,</span><span class="s2">&quot;CSNK1E&quot;</span><span class="p">])]</span>
<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="c1">#AP</span>
    <span class="n">gene_AP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_AP&quot;</span>
    <span class="k">if</span> <span class="n">gene_AP</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="k">elif</span> <span class="n">gene</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;ARNT&quot;</span><span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;ARNT&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">promoter</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MP&quot;</span><span class="p">,</span><span class="s2">&quot;AP&quot;</span><span class="p">]:</span>
            <span class="n">gene_promoter</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">protospacer_1</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_1</span><span class="o">=</span><span class="n">protospacer_1</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

            <span class="n">gene_promoter</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">protospacer_2</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_2</span><span class="o">=</span><span class="n">protospacer_2</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

            <span class="n">gene_promoter</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="n">randrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list_recursive</span><span class="p">))]</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">protospacer_1_random</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_1_random</span><span class="o">=</span><span class="n">protospacer_1_random</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

            <span class="n">gene_promoter</span><span class="o">=</span><span class="n">gene_list_recursive</span><span class="p">[</span><span class="n">randrange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_list_recursive</span><span class="p">))]</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="n">promoter</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">protospacer_2_random</span><span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_promoter</span><span class="p">,</span><span class="n">pval_cutoff</span><span class="o">=</span><span class="n">pvalue</span><span class="p">)</span>
            <span class="n">protospacer_2_random</span><span class="o">=</span><span class="n">protospacer_2_random</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>


            <span class="n">overlap</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protospacer_1</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protospacer_2</span><span class="p">]</span>
            <span class="n">overlap_random_1</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">overlap</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protospacer_1_random</span><span class="p">]</span>
            <span class="n">overlap_random_2</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">overlap</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">protospacer_2_random</span><span class="p">]</span>
            <span class="c1">#if overlap protospacer_1 protospacer_2 overlap_random_2 overlap_random_1 = 0 set to 1</span>
            <span class="c1"># print(protospacer_1,protospacer_2,overlap,overlap_random_1,overlap_random_2)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">overlap_1</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_1</span><span class="p">)</span>
                <span class="n">overlap_2</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_2</span><span class="p">)</span>
                <span class="n">overlap_random_1</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">overlap_random_1</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_1_random</span><span class="p">)</span>
                <span class="n">overlap_random_2</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">overlap_random_2</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">protospacer_2_random</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
                <span class="n">overlap_1</span><span class="o">=</span><span class="mi">0</span>
                <span class="n">overlap_2</span><span class="o">=</span><span class="mi">0</span>
                <span class="n">overlap_random_1</span><span class="o">=</span><span class="mi">0</span>
                <span class="n">overlap_random_2</span><span class="o">=</span><span class="mi">0</span>

            <span class="n">gene_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([[</span><span class="n">gene</span><span class="p">,</span><span class="n">promoter</span><span class="p">,</span><span class="n">overlap_1</span><span class="p">,</span><span class="n">overlap_2</span><span class="p">,</span><span class="n">overlap_random_1</span><span class="p">,</span><span class="n">overlap_random_2</span><span class="p">]])</span>
            <span class="c1"># gene_df_details=pd.DataFrame([[gene,promoter,protospacer_1,protospacer_2,overlap,len(protospacer_1),len(protospacer_2),len(overlap)]])</span>

            <span class="c1"># print(gene_df.shape)</span>
            <span class="n">gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_df</span><span class="p">)</span>
            <span class="c1"># break</span>


<span class="c1"># col_names=[&quot;Gene&quot;,&quot;Promoter&quot;,&quot;Protospacer_1&quot;,&quot;Protospacer_2&quot;,&quot;Overlap&quot;,&quot;Protospacer_Random_1&quot;,&quot;Protospacer_Random_2&quot;]</span>
<span class="c1">#add col_names</span>
<span class="n">col_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span><span class="s2">&quot;Promoter&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_1&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_2&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_Random_1&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_Random_2&quot;</span><span class="p">]</span>
<span class="nb">all</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">gene_list</span><span class="p">))</span>
<span class="nb">all</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">col_names</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AHCYL1
SP1
TGIF1
YWHAZ
P4HB
NCOR2
JARID2
ESR1
PA2G4
GPBP1
DHX30
SAFB
NFE2L2
CHD8
SIN3A
PSMC5
RC3H2
ZNF3
GTF2F1
ADAR
NR2F2
BZW1
FHL2
SET
MYBBP1A
CUX1
NBN
CALR
APP
STAG2
PKM
BRIP1
GATA3
RBM47
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get the tests between two unpaired parametric groups</span>
<span class="c1">#melt for random</span>
<span class="n">all_melt</span><span class="o">=</span><span class="nb">all</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">,</span><span class="s2">&quot;Promoter&quot;</span><span class="p">])</span>
<span class="nb">all</span><span class="o">.</span><span class="n">index</span><span class="o">=</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Gene&quot;</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;_&quot;</span><span class="o">+</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Promoter&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_1&quot;</span><span class="p">],</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_2&quot;</span><span class="p">],</span><span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_Random_1&quot;</span><span class="p">],</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_Random_2&quot;</span><span class="p">],</span><span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_1&quot;</span><span class="p">],</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_Random_1&quot;</span><span class="p">],</span><span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">stats</span><span class="o">.</span><span class="n">ttest_ind</span><span class="p">(</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_2&quot;</span><span class="p">],</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_Random_2&quot;</span><span class="p">],</span><span class="n">equal_var</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
<br/><br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
TtestResult(statistic=-1.200758436011414, pvalue=0.23196364293947955, df=133.97197243239725)
TtestResult(statistic=0.14299988815177855, pvalue=0.8865064991701896, df=133.1170801399246)
TtestResult(statistic=10.31768853968277, pvalue=6.501330271631454e-18, df=111.19538835724522)
TtestResult(statistic=11.997174432050567, pvalue=2.1194060394826254e-21, df=104.97349430144247)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#plot the overlap between the two guides</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">style</span><span class="o">=</span><span class="s2">&quot;white&quot;</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set_context</span><span class="p">(</span><span class="s2">&quot;talk&quot;</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">new_pallete</span><span class="o">=</span><span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">([</span><span class="s2">&quot;#B4D88C&quot;</span><span class="p">,</span><span class="s2">&quot;#C67B6A&quot;</span><span class="p">,</span><span class="s2">&quot;#53449B&quot;</span><span class="p">,</span><span class="s2">&quot;#8C1F1C&quot;</span><span class="p">])</span>
<span class="c1">#show the percentage over &quot;Overlap_length&quot;/&quot;Protospacer_1_length&quot;</span>
<span class="n">sns</span><span class="o">.</span><span class="n">kdeplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">all_melt</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;value&quot;</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="s2">&quot;variable&quot;</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">new_pallete</span><span class="p">,</span>
<span class="p">)</span>
<span class="c1">#save plot</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;figures/overlap_gene_signature.pdf&quot;</span><span class="p">)</span>
<span class="n">gene_list_select</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;guide_id&#39;</span><span class="p">,</span> <span class="n">groups</span><span class="o">=</span><span class="n">gene_list_select</span><span class="o">.</span><span class="n">to_list</span><span class="p">(),</span><span class="n">reference</span><span class="o">=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test&#39;</span><span class="p">)</span>
<span class="n">genelist</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;guide_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="s1">&#39;guide_id&#39;</span><span class="p">,</span><span class="n">groups</span><span class="o">=</span><span class="n">genelist</span><span class="p">,</span> <span class="n">reference</span><span class="o">=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1">#get target gene logfoldchanges</span>
<span class="c1">#get the differentially expressed genes</span>
<span class="n">gene_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">comp_gene_list</span><span class="o">=</span><span class="p">[]</span>

<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">)</span><span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="c1"># print(gene)</span>
    <span class="c1">#AP</span>
    <span class="n">gene_AP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_AP&quot;</span>
    <span class="n">gene_MP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_MP&quot;</span>

    <span class="k">if</span> <span class="n">gene_AP</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">][(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">)</span><span class="o">&amp;</span> <span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">)]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">##form a dataframe with custom colummns</span>
        <span class="n">MP</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_MP</span><span class="p">)</span>
        <span class="n">MP</span><span class="o">=</span><span class="n">MP</span><span class="p">[(</span><span class="n">MP</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">gene</span><span class="p">)]</span>
        <span class="n">AP</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_AP</span><span class="p">)</span>
        <span class="n">AP</span><span class="o">=</span><span class="n">AP</span><span class="p">[(</span><span class="n">AP</span><span class="p">[</span><span class="s2">&quot;names&quot;</span><span class="p">]</span><span class="o">==</span><span class="n">gene</span><span class="p">)]</span>
        <span class="n">promoter_line</span><span class="o">=</span><span class="p">[</span><span class="n">gene</span><span class="p">,</span><span class="n">MP</span><span class="p">[</span><span class="s2">&quot;scores&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">AP</span><span class="p">[</span><span class="s2">&quot;scores&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">MP</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">AP</span><span class="p">[</span><span class="s2">&quot;pvals_adj&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
        <span class="n">comp_gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">promoter_line</span><span class="p">)</span>
        <span class="c1"># break</span>
<span class="c1">#convert to dataframe</span>
<span class="n">comp_gene_list</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">comp_gene_list</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">,</span><span class="s2">&quot;MP_logfoldchange&quot;</span><span class="p">,</span><span class="s2">&quot;AP_logfoldchange&quot;</span><span class="p">,</span><span class="s2">&quot;MP_pvalue&quot;</span><span class="p">,</span><span class="s2">&quot;AP_pvalue&quot;</span><span class="p">])</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="n">gene_list</span><span class="o">=</span><span class="p">[]</span>
<span class="n">comp_gene_list</span><span class="o">=</span><span class="p">[]</span>

<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;perturbation&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
    <span class="c1">#AP</span>
    <span class="n">gene_AP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_AP&quot;</span>
    <span class="n">gene_MP</span><span class="o">=</span><span class="n">gene</span><span class="o">+</span><span class="s2">&quot;_MP&quot;</span>

    <span class="k">if</span> <span class="n">gene_AP</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;guide_id&quot;</span><span class="p">]</span><span class="o">!=</span><span class="s2">&quot;non-targeting_Control&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="k">continue</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1">##form a dataframe with custom colummns</span>
        <span class="n">MP</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_MP</span><span class="p">,</span> <span class="n">pval_cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
        <span class="c1"># MP = MP[MP[&quot;logfoldchanges&quot;].abs()&gt;0.1]</span>
        <span class="n">MP_names</span><span class="o">=</span><span class="n">MP</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="c1"># ,   pval_cutoff=0.05)[&#39;names&#39;]</span>
        <span class="c1"># sc.tl.rank_genes_groups(adata_ideal_subset, &#39;guide_identity&#39;, groups=[&#39;MYBBP1A_AP&#39;], reference=&#39;Negative_Control_Control&#39;, method=&#39;wilcoxon&#39;)</span>
        <span class="n">AP</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">get</span><span class="o">.</span><span class="n">rank_genes_groups_df</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">group</span><span class="o">=</span><span class="n">gene_AP</span><span class="p">,</span> <span class="n">pval_cutoff</span><span class="o">=</span><span class="mf">0.05</span><span class="p">)</span>
        <span class="c1"># AP = AP[AP[&quot;logfoldchanges&quot;].abs()&gt;0.5]</span>

        <span class="n">AP_names</span><span class="o">=</span><span class="n">AP</span><span class="p">[</span><span class="s1">&#39;names&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="c1"># , pval_cutoff=0.05, log2fc_min=0)[&#39;names&#39;]</span>
        <span class="n">overlap</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">MP_names</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">AP_names</span><span class="p">]</span>
        <span class="c1">#get average log2fc for MP, AP and overlap</span>

        <span class="n">gene_df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([[</span><span class="n">gene</span><span class="p">,</span><span class="n">MP_names</span><span class="p">,</span><span class="n">AP_names</span><span class="p">,</span><span class="n">overlap</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">MP_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">AP_names</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">overlap</span><span class="p">)]])</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">gene_df</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
        <span class="n">gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_df</span><span class="p">)</span>


        <span class="c1">##add to another df which includes the gene and their direction</span>
        <span class="n">MP</span><span class="p">[</span><span class="s2">&quot;Guide&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">gene_MP</span>
        <span class="n">comp_gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">MP</span><span class="p">)</span>
        <span class="n">AP</span><span class="p">[</span><span class="s2">&quot;Guide&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">gene_AP</span>
        <span class="n">comp_gene_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">AP</span><span class="p">)</span>
        <span class="c1"># break</span>

<span class="n">col_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">,</span><span class="s2">&quot;MP_Genes&quot;</span><span class="p">,</span><span class="s2">&quot;AP_Genes&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_Genes&quot;</span><span class="p">,</span><span class="s2">&quot;MP_Gene_Num&quot;</span><span class="p">,</span><span class="s2">&quot;AP_Gene_Num&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_Gene_Num&quot;</span><span class="p">]</span>
<span class="nb">all</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">gene_list</span><span class="p">))</span>
<span class="nb">all</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">col_names</span>

<span class="c1">#only filter for genes with successfulKD</span>
<span class="n">all_filt</span><span class="o">=</span><span class="nb">all</span><span class="p">[</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">][</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;successfulKD&quot;</span><span class="p">]</span><span class="o">==</span><span class="s2">&quot;True&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">)]</span>
<span class="n">all_filt</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/singlecell_shortread_analysis/differential_exp_list.csv&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_comp</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">comp_gene_list</span><span class="p">))</span>
<span class="n">all_comp</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="n">MP</span><span class="o">.</span><span class="n">columns</span>
<span class="n">all_comp</span><span class="p">[</span><span class="s2">&quot;Promoter&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">all_comp</span><span class="p">[</span><span class="s2">&quot;Guide&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">all_comp</span><span class="p">[</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">]</span><span class="o">=</span><span class="n">all_comp</span><span class="p">[</span><span class="s2">&quot;Guide&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">all_comp_mean</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;logfoldchanges&#39;</span><span class="p">:</span> <span class="s1">&#39;mean&#39;</span><span class="p">})</span>
<span class="n">all_comp_mean</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">all_comp_mean</span><span class="o">.</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">,</span><span class="s2">&quot;logfoldchanges_mean&quot;</span><span class="p">]</span>
<span class="n">all_comp</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">all_comp_mean</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">)</span>
<span class="n">all_comp_mean</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span><span class="s1">&#39;Target_Gene&#39;</span><span class="p">:</span><span class="s2">&quot;count&quot;</span><span class="p">})</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">:</span><span class="s2">&quot;Count&quot;</span><span class="p">})</span>
<span class="n">all_comp_mean</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">all_comp</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">all_comp_mean</span><span class="p">,</span><span class="n">on</span><span class="o">=</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">)</span>
<span class="n">all_comp</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s2">&quot;Guide&quot;</span><span class="p">)</span>
<span class="n">all_comp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">loc</span><span class="o">+</span><span class="s2">&quot;files/singlecell_shortread_analysis/differential_exp_list_full.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Mutual_Exclusivity&quot;</span><span class="p">]</span><span class="o">=</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;Overlap_Gene_Num&quot;</span><span class="p">]</span><span class="o">/</span><span class="p">(</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;MP_Gene_Num&quot;</span><span class="p">]</span><span class="o">+</span><span class="nb">all</span><span class="p">[</span><span class="s2">&quot;AP_Gene_Num&quot;</span><span class="p">])</span>

<span class="n">n_term</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s2">&quot;alt-prom-crispr-fiveprime/reference/all_pivot_simple_nterm.txt&quot;</span><span class="p">)</span>
<span class="n">n_term</span><span class="o">=</span><span class="nb">all</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_term</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;Gene_symbol&quot;</span><span class="p">,</span><span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">)</span>

<span class="n">n_term</span><span class="o">=</span><span class="n">n_term</span><span class="p">[[</span><span class="s2">&quot;MP_Gene_Num&quot;</span><span class="p">,</span><span class="s2">&quot;AP_Gene_Num&quot;</span><span class="p">,</span><span class="s2">&quot;Overlap_Gene_Num&quot;</span><span class="p">,</span><span class="s2">&quot;Mutual_Exclusivity&quot;</span><span class="p">,</span><span class="s2">&quot;Ensembl_ID&quot;</span><span class="p">,</span><span class="s2">&quot;Gene_symbol&quot;</span><span class="p">,</span><span class="s2">&quot;Nterminus_Change&quot;</span><span class="p">]]</span>
<span class="c1"># all_comp_pivot</span>
<span class="n">all_comp_all</span><span class="o">=</span><span class="n">all_comp</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">n_term</span><span class="p">,</span> <span class="n">left_on</span><span class="o">=</span><span class="s2">&quot;Target_Gene&quot;</span><span class="p">,</span> <span class="n">right_on</span><span class="o">=</span><span class="s2">&quot;Gene_symbol&quot;</span> <span class="p">,</span><span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">)</span>
<span class="n">all_comp_all</span><span class="o">=</span><span class="n">all_comp_all</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;Nterminus_Change&quot;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="7A_cellphase_model.html" class="btn btn-neutral float-left" title="Guide Calling Using the Cellranger Module HDF5 File" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="9A_cnv_score.html" class="btn btn-neutral float-right" title="Guide Calling Using the Cellranger Module HDF5 File" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2026, Helen King.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>